lnrpc: Add destination output information (pkScript, output index, amount and if output belongs to the node)

bjarnemagnussen opened the pull request:
Fixes #2835.

This PR adds destination output information to the `TransactionDetail` structure returned by both RPC `GetTransactions` and `SubscribeTransactions`. 

The additional fields include the output script (`PkScript`), amount (`Amount`), index (`OutputIndex`) and a bool (`IsOurAddress`) to denote if the output belongs to the node's internal wallet or not.

Since all of this information is already available when constructing the `TransactionDetail`, this information can just be added as it can be useful in many applications and depending on the use case may also solve issue #4826.

#### Note
The output index does not simply correspond to the position in the `DestOutput` array, as unrecognized outputs are skipped:
https://github.com/lightningnetwork/lnd/blob/756a0ab39240dcfd15dea325c5921ef84e7109d9/lnwallet/btcwallet/btcwallet.go#L927-L935

guggero commented on lnwallet/interface.go:
```diff
@@ -80,6 +80,14 @@ type Utxo struct {
 }
 
+// DestOutput contains additional information on a destination address.
+type DestOutput struct {
```
Should we call this struc `OutputDetail`, add the `Address btcutil.Address` field to it and completely replace the `DestAddresses` below? Could call the field `OutputDetails` instead of `DestOutputs` too.

bjarnemagnussen replied:
I agree and have renamed it!

guggero commented on lnrpc/rpc.proto:
```diff
@@ -537,6 +537,20 @@ message Utxo {
     int64 confirmations = 6;
 }
 
+message DestOutput {
```
Same here, could just be `OutputDetail` and include the address (as a string). Then we can mark the `dest_addresses` as deprecated (but still keep the field for the moment to remain backward compatible).

bjarnemagnussen replied:
I have added the destination address to the `OutputDetail` structure. However, I do have a small reservation regarding how the address is calculated, see my other comment for the updated code.

guggero reviewed (changes requested):
Great PR, thank you! This additional info should be very useful.

I think we should merge the new info with the address internally and deprecate the address field in the RPC. But other than that looks great to me!

guggero commented:
> The output index does not simply correspond to the position in the DestOutput array, as unrecognized outputs are skipped:

Perhaps we could even optimize the output here. If we can't parse the address, we set the address to nil (and empty string on the RPC) but still include the other details. That way we could get rid of that `continue` statement you linked.

bjarnemagnussen commented:
> > The output index does not simply correspond to the position in the DestOutput array, as unrecognized outputs are skipped:
> 
> Perhaps we could even optimize the output here. If we can't parse the address, we set the address to nil (and empty string on the RPC) but still include the other details. That way we could get rid of that `continue` statement you linked.

I agree and think this makes a lot of sense. My only reservation is in the way that the address is calculated. As it currently is, the returned address for a single output script is a slice of potential multiple addresses. Normally the slice will just contain a single valid address that corresponds to exactly the output script. However, in the extremely rare case of a MultiSig script _NOT_ encoded via a P2SH, the returned slice of addresses will simply contain a list of public key addresses from the multisig script (since the direct multisig script does not have an address representation).

I have therefore added some logic to make sure in this case we return an empty address.

guggero commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -923,18 +923,44 @@ func minedTransactionsToDetails(
+				// Use the supported address only if it encodes
+				// a single valid address corresponding to the
+				// output script.
+				if len(outAddresses) == 1 && sc != txscript.MultiSigTy {
```
Ah, I wasn't aware of this edge case. I wonder if we should just make the `Address` field of `OutputDetail` a slice and add all addresses to it in this case. Then we could get rid of the `DestAddresses` field here (and update any call sites that use the `TransactionDetails`)...

And on the RPC level the deprecated addresses slice would just be a list of all of these. Not sure about turning the `lnrpc.OutputDetail.address` into a slice either, seems a bit overkill if it only contains one address in almost any case. We could still return it empty if there's more than one address in the internal struct. Then an RPC user can still parse the PK script. What do you think?

guggero replied:
Just trying to keep this internal wallet code as "clean" as possible and move any special cases to the RPC server.

bjarnemagnussen replied:
> Ah, I wasn't aware of this edge case. I wonder if we should just make the Address field of OutputDetail a slice and add all addresses to it in this case. Then we could get rid of the DestAddresses field here (and update any call sites that use the TransactionDetails)...

I will update the code to do that!

> And on the RPC level the deprecated addresses slice would just be a list of all of these. Not sure about turning the lnrpc.OutputDetail.address into a slice either, seems a bit overkill if it only contains one address in almost any case. We could still return it empty if there's more than one address in the internal struct. Then an RPC user can still parse the PK script. What do you think?

I think this would also be acceptable. Now the only way that a MultiSig output script would still be represented by a (wrong) P2PKH address is if it is a 1-of-1 multisig. But I think for such an edge case that is OK, since I don't see how this could be abused as the MultiSig script must be 1-of-1 and cannot contain some additional spending paths, right? 

Additionally, I was thinking about adding the address type as a field to the `OutputDetail` structure, just as for `Utxso` and returned when requesting unspents?

guggero replied:
> if it is a 1-of-1 multisig. But I think for such an edge case that is OK

I agree.

> I was thinking about adding the address type as a field to the OutputDetail structure

That could be useful, yes. But you're thinking about the type returned from `txscript.ExtractPkScriptAddrs()`, right? For that we'd probably need a new enum on the RPC level because the existing `AddressType` only knows P2WKH and NP2WKH (and also has a slightly different context than what we're talking about here).

If you're willing to add that, please go ahead :) My naming suggestion for the enum would be `OutputScriptType` or something like that.

bjarnemagnussen commented:
I have squashed the new changes:

 * Remove `DestAddresses` from the `lnwallet.TransactionDetail` structure as it is otherwise redundantly available via `OutputDetails` and just re-populated to the RPC for backwards compatibility.
 * Introduced another field to `OutputDetail` called `OutputType` that will be of `OutputScriptType` to denote the kind of output script. However, I am not sure if alternatively defining it as a `string` and using the `String()` method on a `ScriptClass` is the more pragmatic solution? (See the definition [here](https://github.com/btcsuite/btcd/blob/86a17263b0ff614cd9e1fb3ef4d2757af00e8578/txscript/standard.go#L64-L86))

guggero commented on lnrpc/rpc.proto:
```diff
@@ -537,6 +537,38 @@ message Utxo {
+    SCRIPT_HASH = 1;
+    WITNESS_V0_PUBKEY_HASH = 2;
+    WITNESS_V0_SCRIPT_HASH = 3;
+    PUBKEY = 4;
```
We're planning on upgrading the `grpc` library version soon. There have been some recent changes which will warn if enum value names inside the same gRPC package collide. So if we'd add another enum with a different name but the same `PUBKEY` value name for example, we'd get a nasty error. That's why I think we should start adding a prefix to all our (new) enums.
So something like this:

```
enum OutputScriptType {
    SCRIPT_TYPE_PUBKEY_HASH = 0;
    SCRIPT_TYPE_SCRIPT_HASH = 1;
    SCRIPT_TYPE_WITNESS_V0_PUBKEY_HASH = 2;
    SCRIPT_TYPE_WITNESS_V0_SCRIPT_HASH = 3;
    SCRIPT_TYPE_PUBKEY = 4;
...
```

I know that makes it impossibly long in the generated golang code. But migrating enum names because of a collision is even more tedious.

bjarnemagnussen replied:
Makes sense! I have changed it accordingly.

guggero commented on lnrpc/rpc_utils.go:
```diff
@@ -17,8 +17,28 @@ func RPCTransactionDetails(txns []*lnwallet.TransactionDetail) *TransactionDetai
+		// Re-package destination output information.
+		var outputDetails []*OutputDetail
+		for _, o := range tx.OutputDetails {
+			// Note: DestAddresses is deprecated but we keep populating it
```
nit: lines too long here and in the next commit.

bjarnemagnussen replied:
Ah yes, I may have had the standard 4-whitespace settting on. I have changed it accordingly and tried to make it more readable instead of many breaks.

guggero reviewed (commented):
Looks good, thanks for adding the new enum. Just one comment about the enum values and a few nits, other than that I think this is ready.

guggero reviewed (approved):
Thank you for the super quick updates, LGTM :tada:

lightninglabs-deploy commented:
@ellemouton: review reminder

ellemouton commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -923,18 +923,37 @@ func minedTransactionsToDetails(
+				OutputIndex:  i,
+				Value:        btcutil.Amount(txOut.Value),
+				IsOurAddress: isOurAddress[i],
+			})
```
not sure of this. If `err != nil` then we probably dont want to include the OutputDetail either, right? also, if err != nil then for the above, the output detail added will have a nil output type and nil addresses slice. So i guess my main question is, if we do not support an address, do we still want to create an output detail entry for it? and if so, should we add an "Unsupported" field to the Output detail or something?"

ellemouton replied:
ah ok, i see that the `ExtractPkScriptAddrs` func will return a `NonStandard` script class if an error is returned. So this should be g2g 👍

ellemouton commented on lnrpc/rpc.proto:
```diff
@@ -537,6 +537,38 @@ message Utxo {
+    OutputScriptType output_type = 1;
+
+    // The address
+    string address = 2;
```
in the first commit, the OutputDetail struct has an array of addresses instead of just 1. I assumed that is for a multisig script or something? in anycase, if we do a slice there, then this should also be be `repeated`. otherwise, the struct should be changed to just have 1 addr too.

bjarnemagnussen replied:
The reason to use a slice for it was to deprecate `DestAddresses` but still expose it on the RPC for backwards compatibility. However, we can also re-introduce the `DestAddresses` directly to the interface and have the `OutputDetail` only contain the single destination address if available?

Note that the `DestAddresses` field is a bit weird, because if an output is not re-presentable by an address (e.g. due to `NonStandard` type or the `MultiSigType`), then the indices in the `DestAddresses` do not correspond to the outputs.

The reason that `MultiSigType` returns a slice of addresses, is because this is the case where the multisig script is directly put inside the scriptPubKey, hence not behind a hash. Therefore it cannot be represented by an address. It is questionable if a slice of addresses for the pubkeys referred to in the multisig script is helpful though. :)

ellemouton commented on lnrpc/rpc_utils.go:
```diff
@@ -17,8 +17,30 @@ func RPCTransactionDetails(txns []*lnwallet.TransactionDetail) *TransactionDetai
+
+			var address string
+			if len(o.Addresses) == 1 {
+				address = o.Addresses[0].EncodeAddress()
```
yeah so following up on my comment from before, what about the other addresses here?

bjarnemagnussen replied:
The `DestAddresses` field is a bit weird, because if an output is not re-presentable by an address (e.g. due to `NonStandard` type or the `MultiSigType`), then the indices in the `DestAddresses` do not correspond to the outputs.

The reason that `MultiSigType` returns a slice of addresses, is because this is the case where the multisig script is directly put inside the scriptPubKey, hence not behind a hash. Therefore it cannot be represented by an address. It is questionable if a slice of addresses for the pubkeys referred to in the multisig script is helpful though. :)

ellemouton replied:
ok cool, chatted to oli about this and he agrees with you 👍

ellemouton reviewed (commented):
left some comments. Also, needs a rebase :)

bjarnemagnussen commented:
Thanks @ellemouton for your comments! I have rebased this PR. Also I have tried to clarify the weird behaviour of the legacy `DestAddresses` field in a comment and am open to handle it differently.

ellemouton reviewed (approved):
tACK 🚀 

i think it needs another rebase @bjarnemagnussen

bjarnemagnussen commented:
Thanks @ellemouton! I have rebased and added release notes!
